src/compiler/moduleSpecifiers.ts

```diff
@@ -687,13 +687,17 @@ namespace ts.moduleSpecifiers {
             case Ending.Index:
                 return noExtension;
             case Ending.JsExtension:
-                return noExtension + getJSExtensionForFile(fileName, options);
+                return noExtension + getJSExtensionForFileIfApplicable(fileName, options);
             default:
                 return Debug.assertNever(ending);
         }
     }
 
-    function getJSExtensionForFile(fileName: string, options: CompilerOptions): Extension {
+    function getJSExtensionForFileIfApplicable(fileName: string, options: CompilerOptions): Extension {
+        return getJSExtensionForFile(fileName, options) ?? Debug.fail(`Extension ${extensionFromPath(fileName)} is unsupported:: FileName:: ${fileName}`);
+    }
+
+    export function getJSExtensionForFile(fileName: string, options: CompilerOptions): Extension | undefined {
         const ext = extensionFromPath(fileName);

```

Comment:
I thought it was sketchy that this function didn’t return `undefined`, so I looked and found that it too has an undefined-returning counterpart, and this one throws! This may need to be swapped for `tryGetExtensionFromPath` for good measure. This also makes me realize we should have a test that actually returns completions for a non-source-file file extension, like `index.css` or something.

Reply:
I don't think we even read other extension types from disk, but I think you mean something like

```ts
//@filename: index.css
body {}

//@filename: index.html
<!DOCTYPE html>

//@filename: foo.ts
import "./
```

And expect nothing at the uncompleted string?

Reply:
Hm, I looked at stringCompletions and saw `tryReadDirectory`, so I assumed we were offering everything we could find, with extension substitutions as appropriate. Not sure what’s going on then 🤔

Reply:
You could certainly be seeing an artifact of the fourslash client 😬. Have you checked the behavior against a Real Editor?

Reply:
I was pretty sure that `tryReadDirectory` only returned files that matched its `extensions` argument, if provided. I might be misreading though.

Reply:
either way, updated to not use the throwing variant :)

Reply:
Ah, I missed that parameter 🤦‍♂️ 

It might actually be nice to allow all extensions here in the future, since we can no longer seriously claim to believe we’re in a world where people only import things we know how to analyze. This doesn’t even account for `declare module "*.css"` pattern ambient module declarations. But that’s for another day. Thanks!

Reply:
Oh geez TIL. Is that a [wildcard module declaration](https://www.typescriptlang.org/docs/handbook/modules.html#wildcard-module-declarations)?